### PR TITLE
Refactor and improve the C++ JitBuilder API generator and associated tests

### PR DESCRIPTION
**Pull request by Leonardo2718**
The following improvements are done the the C++ API generator:

- Split the generator tests logically across multiple files for easier maintenance
- Refactor the generator into a class for better encapsulation
- Add some new helpers that will be useful for creating a C API generator

### CODE REVIEW

**Comment by Leonardo2718** [comment]
@genie-omr build all

File: jitbuilder/apigen/cppgen.py [hunk 1]
```diff
@@ -62,878 +96,861 @@
...
-    if desc.is_vararg():
+        # include headers for each defined class
+        for c in api_desc.get_class_names():
+            writer.write(self.generate_include(c + ".hpp"))
         writer.write("\n")
-        write_vararg_service_impl(writer, desc, class_name)
 
-def write_common_impl(writer, api_desc):
-    """Writes the implementation of all client API (non-class) services."""
+        # write delcarations for all services
```
Comment (mstoodle) [inline jitbuilder/apigen/cppgen.py]: typo

File: jitbuilder/apigen/cppgen.py [hunk 2]
```diff
@@ -62,878 +96,861 @@
...
+            writer.write("return clientObj;\n")
+        else:
+            writer.write("auto ret = " + impl_call + ";\n")
+            for parm in desc.parameters():
+                self.write_arg_return(writer, parm)
+            writer.write("return ret;\n")
+
+        writer.write("}\n")
 
-    write_allocators_setter(writer, api
```
Comment (mstoodle) [inline jitbuilder/apigen/cppgen.py]: typo: opened

**Comment by Leonardo2718** [comment]
@mstoodle fixed those couple of typos. :slightly_smiling_face:

> I didn't find these commits to be particularly easy to review, mostly due to the large blocks of stuff that changed.

Yeah, I'm not supper happy with how git shows the diff for that commit. Actually doing the change was much easier than what the diff makes it look like (just indent the code and add `self` to a bunch of places). Still thought it best to have those tests though, just to be sure :wink:. Hopefully, follow up refactoring will be easier to review.

@genie-omr build all

**Comment by mstoodle** [comment]
@Leonardo2718 are you aware of any other reasons this should not be merged yet? Putting on my OpenJ9 hat in the wrong place (but just for a moment!), have you tried an openj9 build to make sure this won't break it? :)

**Comment by mstoodle** [comment]
Note that I just asked a question about whether you have done it which should not be taken to suggest that you should have done it or that you should do it if you haven't...

**Comment by Leonardo2718** [comment]
I don't think there's anything in this change specifically that would cause OpenJ9 to break. If there is, I would have expected the existing code to also break. That said, I have not tried doing an OpenJ9 build.

**Comment by mstoodle** [comment]
sorry; I think I had another PR in my brain when I asked this question. Your answer is completely reasonable.

**Pull request merged**